### PR TITLE
Feat/remove protocol params default

### LINKED ISSUE #1876: Should `ProtocolParameters` impl `Default`?
## Description

We should think about whether that really makes sense considering:
* values change over time through fluid protocol upgrades (what happens then? Updating the defaults?)
* there are several networks in parallel (not clear why defaulting to a particular one of those?)
* buggy as shown with https://github.com/iotaledger/iota-sdk/issues/1808

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

Removes the default impls for protocol params and the inner structures, and replaces them with the `iota_mainnet_v3_protocol_parameters` function for testing. This was heavily dependent on packable changes so those were made in https://github.com/iotaledger/common-rs/pull/76. Additionally, the `NetworkInfo` struct has been made optional on the client because the values cannot be defaulted.

## Links to any relevant issues

Closes #1876

### CODE REVIEW

File: sdk/src/client/core.rs [hunk 1]
```diff
@@ -107,24 +108,45 @@ impl ClientInner {
...
-                if current_time < last_sync {
-                    return Ok(self.network_info.read().await.clone());
-                }
+            self.fetch_network_info().await
+        }
+
+        #[cfg(not(target_family = "wasm"))]
+        Ok(if let Some(info) = &*self.network_info.read().await {
+            info.clone()
+        } else {
```
Comment (Thoralf-M) [inline sdk/src/client/core.rs]: This should have the timestamp logic from fetch_network_info, otherwise we will never update the network info from the node

File: sdk/src/client/node_manager/syncing.rs [hunk 2]
```diff
@@ -110,14 +111,21 @@ impl ClientInner {
...
                     .clone();
+                if let Some(network_info) = network_info.as_mut() {
+                    network_info.tangle_time = info.status.relative_accepted_tangle_time;
+                    network_info.protocol_parameters = protocol_parameters;
+                } else {
+                    network_info.replace(NetworkInfo {
+                        protocol_parameters,
+                        tangle_time: info.status.relative_accepted_tangle_time,
+                    });
+                }
```
Comment (Thoralf-M) [inline sdk/src/client/node_manager/syncing.rs]: Does anything speak against just always replacing it?

File: sdk/src/client/core.rs (hunk 1 above)
Comment (DaughterOfMars) [reply sdk/src/client/core.rs]: That's what it did before, and relied on sync to update it

File: sdk/src/client/node_manager/syncing.rs (hunk 2 above)
Comment (DaughterOfMars) [reply sdk/src/client/node_manager/syncing.rs]: no, I don't know why I did this lol

File: sdk/src/client/core.rs (hunk 1 above)
Comment (Thoralf-M) [reply sdk/src/client/core.rs]: Ah nvm I thought this is the syncing code

File: sdk/src/types/block/payload/signed_transaction/transaction.rs [hunk 3]
```diff
@@ -328,16 +328,14 @@ impl WorkScore for Transaction {
     }
 }
 
-fn verify_network_id<const VERIFY: bool>(network_id: &u64, visitor: &ProtocolParameters) -> Result<(), Error> {
-    if VERIFY {
```
Comment (Thoralf-M) [inline sdk/src/types/block/payload/signed_transaction/transaction.rs:332]: Just checking, visitor doesn't need to be made optional and then a check if it's Some()?

Comment (DaughterOfMars) [reply sdk/src/types/block/payload/signed_transaction/transaction.rs:332]: No, because these are only called if it is `Some`

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/core/Cargo.toml [hunk 4]
```diff
@@ -11,6 +11,7 @@ publish = false
 [dependencies]
 iota-sdk = { path = "../../sdk", default-features = false, features = [
     "wallet",
+    "samples",
```
Comment (thibault-martinez) [inline bindings/core/Cargo.toml]: Is this really a good name? Not very descriptive of what it contains tbh

File: sdk/src/client/node_manager/syncing.rs [hunk 5]
```diff
@@ -3,6 +3,7 @@
 
 #[cfg(not(target_family = "wasm"))]
 use {
+    crate::client::NetworkInfo,
     crate::types::block::PROTOCOL_VERSION,
```
Comment (thibault-martinez) [inline sdk/src/client/node_manager/syncing.rs]: merge crate imports

File: sdk/src/types/block/context_input/mod.rs [hunk 6]
```diff
@@ -87,7 +87,7 @@ pub(crate) type ContextInputCount =
 #[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Deref, Packable)]
 #[packable(unpack_error = Error, with = |e| e.unwrap_item_err_or_else(|p| Error::InvalidContextInputCount(p.into())))]
 pub struct ContextInputs(
-    #[packable(verify_with = verify_context_inputs_packable)] BoxedSlicePrefix<ContextInput, ContextInputCount>,
+    #[packable(verify_with = verify_context_inputs)] BoxedSlicePrefix<ContextInput, ContextInputCount>,
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/mod.rs:90]: Just making sure, since `ContextInputs` doesn't have a packable `unpack_visitor` attribute, it'll still be forwarded a `()` and then calling the verify if the parent type does have a visitor, right?

File: bindings/core/Cargo.toml (hunk 4 above)
Comment (DaughterOfMars) [reply bindings/core/Cargo.toml]: suggestions?

File: sdk/src/types/block/context_input/mod.rs (hunk 6 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/context_input/mod.rs:90]: Basically, yes

File: sdk/src/client/builder.rs [hunk 7]
```diff
@@ -275,11 +292,10 @@ impl ClientBuilder {
 
 /// Struct containing network related information
 // TODO do we really want a default?
```
Comment (Thoralf-M) [inline sdk/src/client/builder.rs]: Can be removed now?
```suggestion
```

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/tests/client/signing/account.rs [hunk 8]
```diff
@@ -202,6 +206,7 @@ async fn account_reference_unlocks() -> Result<()> {
         )
         .with_outputs(outputs)
         .with_creation_slot(slot_index + 1)
+        .with_capabilities([TransactionCapabilityFlag::BurnMana])
```
Comment (thibault-martinez) [inline sdk/tests/client/signing/account.rs:209]: Why are we adding that?

Comment (Thoralf-M) [reply sdk/tests/client/signing/account.rs:209]: It was added before automatic mana allotment so it doesn't fail because of mana remainder, but now we can remove it

Comment (thibault-martinez) [reply sdk/tests/client/signing/account.rs:209]: I tried removing them and tests are failing

Comment (thibault-martinez) [reply sdk/tests/client/signing/account.rs:209]: But since it's signing tests, maybe it's not that bad to keep them since it's not really the point of the test

Comment (Thoralf-M) [reply sdk/tests/client/signing/account.rs:209]: Yeah one would need to adjust them if it should pass without it

File: sdk/src/client/core.rs [hunk 9]
```diff
@@ -46,7 +46,7 @@ impl core::ops::Deref for Client {
 pub struct ClientInner {
     /// Node manager
     pub(crate) node_manager: RwLock<NodeManager>,
-    pub(crate) network_info: RwLock<NetworkInfo>,
+    pub(crate) network_info: RwLock<Option<NetworkInfo>>,
```
Comment (thibault-martinez) [inline sdk/src/client/core.rs]: Should we not request the info from the node if no ProtocolParameters were provided? Sounds a bit annoying to have an Option here. Not being able to have infos from both the user and the node seems unlikely and should error?

Comment (DaughterOfMars) [reply sdk/src/client/core.rs]: I don't disagree, I'll try it

File: sdk/tests/client/input_selection/account_outputs.rs [hunk 10]
```diff
@@ -1809,84 +1809,86 @@ fn min_allot_account_mana_additional() {
...
-    let account_id_2 = AccountId::from_str(ACCOUNT_ID_2).unwrap();
-
-    let additional_allotment = 1000;
-    let txn_required_mana_allotment = 271_000;
-    // The account does not have enough to cover the requirement
-    let account_mana = txn_required_mana_allotment - 100;
-    // But there is additional available mana elsewhere
-    let additional_available_mana = additional_allotment + 111;
+// #[test]
+// fn min_allot_account_mana_cannot_select_additional() {
```
Comment (thibault-martinez) [inline sdk/tests/client/input_selection/account_outputs.rs:1821]: Why?

Comment (DaughterOfMars) [reply sdk/tests/client/input_selection/account_outputs.rs:1821]: Reliant on #2036

Comment (thibault-martinez) [reply sdk/tests/client/input_selection/account_outputs.rs:1821]: Ok, let's not forgot to uncomment/fix it there then, put a comment or resolve this

**Review by Thoralf-M — CHANGES_REQUESTED** [review]
Failing tests

File: sdk/src/client/core.rs [hunk 11]
```diff
@@ -31,6 +31,7 @@ use crate::{
 #[derive(Clone)]
 pub struct Client {
     pub(crate) inner: Arc<ClientInner>,
+    pub(crate) network_info: Arc<RwLock<NetworkInfo>>,
```
Comment (thibault-martinez) [inline sdk/src/client/core.rs:34]: Can you ELI5 why it's been moved to the Client pls?

File: sdk/src/client/node_manager/syncing.rs [hunk 12]
```diff
@@ -69,7 +66,7 @@ impl ClientInner {
 
         log::debug!("sync_nodes");
         let mut healthy_nodes = HashSet::new();
-        let mut network_nodes: HashMap<String, Vec<(ProtocolParameters, Node, Option<u64>)>> = HashMap::new();
+        let mut network_nodes: HashMap<String, Vec<(_, Node, Option<u64>)>> = HashMap::new();
```
Comment (thibault-martinez) [inline sdk/src/client/node_manager/syncing.rs:69]: Just curious, why?

File: sdk/src/wallet/core/builder.rs [hunk 13]
```diff
@@ -247,6 +247,30 @@ where
...
+            .await?;
+
+        // May create a default Ed25519 wallet address if there's a secret manager.
+        let wallet_address = if self.address.is_none() && self.secret_manager.is_some() {
+            let bech32_hrp = client.network_info.read().await.protocol_parameters.bech32_hrp();
+            let address = self.create_default_wallet_address(bech32_hrp).await?;
+            address
+        } else {
+            return Err(crate::wallet::Error::MissingParameter("address"));
+        };
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/builder.rs]: This all looks like the result of a failed conflict resolution?

File: sdk/src/client/core.rs (hunk 11 above)
Comment (DaughterOfMars) [reply sdk/src/client/core.rs:34]: The ClientInner is needed to call network methods, like to fetch the info. But I can't create the ClientInner without the info.

File: sdk/src/client/node_manager/syncing.rs (hunk 12 above)
Comment (DaughterOfMars) [reply sdk/src/client/node_manager/syncing.rs:69]: Just didn't feel like importing it with the feature gate

File: sdk/src/wallet/core/builder.rs (hunk 13 above)
Comment (DaughterOfMars) [reply sdk/src/wallet/core/builder.rs]: I re-added it because it didn't make sense to me that we can't generate an address now

Comment (thibault-martinez) [reply sdk/src/wallet/core/builder.rs]: But `self.address` can't be None at this point

Comment (Thoralf-M) [reply sdk/src/wallet/core/builder.rs]: Why not? If you provide the bip path it should generate it in L237?

Comment (DaughterOfMars) [reply sdk/src/wallet/core/builder.rs]: oh, missed that haha

File: sdk/src/wallet/core/builder.rs [hunk 14]
```diff
@@ -247,6 +247,21 @@ where
...
 
+        // May use a previously stored wallet address if it wasn't provided
+        if self.address.is_none() {
+            self.address = loaded_wallet_builder
+                .as_ref()
+                .and_then(|builder| builder.address.clone());
+        }
+
+        // Create the node client.
+        let client = self
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/builder.rs]: Already done L215

File: sdk/src/wallet/core/builder.rs [hunk 15]
```diff
@@ -247,6 +247,21 @@ where
             self.alias = loaded_wallet_builder.as_ref().and_then(|builder| builder.alias.clone());
         }
 
+        // May use a previously stored wallet address if it wasn't provided
+        if self.address.is_none() {
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/builder.rs]: Already done L184

File: sdk/src/client/node_manager/syncing.rs (hunk 12 above)
Comment (thibault-martinez) [reply sdk/src/client/node_manager/syncing.rs:69]: Fair enough

**Review by thibault-martinez — DISMISSED** [review]

File: bindings/core/src/method/utils.rs [hunk 16]
```diff
@@ -120,13 +154,19 @@ pub enum UtilsMethod {
...
+        output_id: OutputId,
+    },
     /// Computes the slot commitment id from a slot commitment.
     #[serde(rename_all = "camelCase")]
-    ComputeSlotCommitmentId { slot_commitment: SlotCommitment },
+    ComputeSlotCommitmentId {
+        slot_commitment: SlotCommitment,
+    },
     /// Returns the hex representation of the serialized output bytes.
     #[serde(rename_all = "camelCase")]
```
Comment (Thoralf-M) [inline bindings/core/src/method/utils.rs]: Not needed, but also not really a problem

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/core/src/method/utils.rs (hunk 16 above)
Comment (Thoralf-M) [reply bindings/core/src/method/utils.rs]: The `#[serde(rename_all = "camelCase")]`

**Review by thibault-martinez — APPROVED** [review]

**Review by Thoralf-M — DISMISSED** [review]

**Review by Alex6323 — DISMISSED** [review]

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Python tests failing

**Review by thibault-martinez — COMMENTED** [review]
Still failing :trollface:

**Review by thibault-martinez — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]
